### PR TITLE
[FIX] l10n_it_edi: Fix Discount line addition

### PR DESCRIPTION
**Pull request by adda-odoo**
In the current implementation, a new invoice line is created in the cache using the new() as a form, which is then used to populate the invoice line. However, the new() method does not have an `__enter__` attribute since it is not wrapped with `contextmanager`, hence the `with` statement fails.

Solution is to simplify the create using Command and attaching it to the O2M field.

opw-3248135
opw-3278824

### CODE REVIEW

**Comment by ThomasBeckers** [comment]
Hello @adda-odoo
I would probably use the create method method used above in the code 
https://github.com/odoo/odoo/blob/b02ecb8fffe82e7ab0f20f2b01ce11c582113583/addons/l10n_it_edi/models/account_edi_format.py#L697-L700
You also removed the clear of tax on the line that was there before. It will impact tax set on the line created so I would keep it
https://github.com/odoo/odoo/blob/63fd595a71695f90dbe42dc461ee1799f15ed747/addons/l10n_it_edi/models/account_edi_format.py#L726

**Comment by adda-odoo** [comment]
Hey @ThomasBeckers, made the changes!

**Comment by lordkrandel** [comment]
Wait... I remember seeing some "SCONTO" line before... how is that possible?

**Comment by adda-odoo** [comment]
> Wait... I remember seeing some "SCONTO" line before... how is that possible?

Hey @lordkrandel, do you mean elsewhere in the code? This is the block of code that adds the discount move line to the invoice as per the "ScontoMaggiorazione" tag in the xml. Am I missing something here?

**Comment by lordkrandel** [comment]
No, I mean in a previous version, in an exported invoce... So if the method doesn't exist at all... it's strange! O.o
Do we do something different?

**Comment by lordkrandel** [comment]
Ok, we got to the root of it.
The line was created with a Form component.
https://discord.com/channels/@me/871736761309618178/1106522270589014048
Can you add a test for the case of the discounted invoice?
Thank you!

**Comment by adda-odoo** [comment]
Hey @lordkrandel, if I update the test xml file `IT01234567890_FPR01.xml` to contain the discount line, then this method would fail. https://github.com/odoo/odoo/blob/0cff545e91adefe9804d8cb22db1f74726ea64e3/addons/l10n_it_edi/tests/test_edi_import.py#L88
But after the fix, it passes. Is that enough to test the discount? Currently, the test xml file does not contain the 'sconto' tag.

**Comment by lordkrandel** [comment]
No, I'm sorry, please make a duplicated test with a specific XML file please, as for example it's being done in https://github.com/odoo/odoo/blob/master/addons/l10n_it_edi_withholding/tests/test_withholding.py
It's easier for everyone to know which specific test is failing, and also to expand if we change something in the discount.

File: addons/l10n_it_edi/tests/test_edi_import.py [hunk 1]
```diff
@@ -129,3 +131,14 @@ def test_receive_same_vendor_bill_twice(self):
...
+    def test_discount_line(self):
+        """Test discount line addition after bill import"""
+
+        discount_line = {'name':'SCONTO'}
+
+        content = etree.fromstring(self.invoice_discount_content)
+        invoices = self.edi_format._create_invoice_from_xml_tree(self.invoice_discount_file, content)
+
+        invoice_lines = invoices.line_ids
+        self.assertEqual(invoice_lines.name,discount_line['name'])
```
Comment (lordkrandel) [inline addons/l10n_it_edi/tests/test_edi_import.py:144]:
```suggestion
        content = etree.fromstring(self.invoice_discount_content)
        invoices = self.edi_format._create_invoice_from_xml_tree(self.invoice_discount_file, content)
        self.assertEqual(invoices.line_ids.name, 'SCONTO')
```

**Review by lordkrandel — CHANGES_REQUESTED** [review]
Some cosmetics. Think we'll have to squash the commits, no need to create a second just for the test.

**Review by lordkrandel — APPROVED** [review]
LGTM

File: addons/l10n_it_edi/models/account_edi_format.py [hunk 2]
```diff
@@ -722,11 +722,13 @@ def _import_fattura_pa(self, tree, invoice):
...
-                        invoice_line_global_discount.sequence = sequence
-                        invoice_line_global_discount.name = 'SCONTO' if general_discount < 0 else 'MAGGIORAZIONE'
-                        invoice_line_global_discount.price_unit = general_discount
+                    invoice_form.invoice_line_ids.create({
+                        'move_id': invoice_form.id,
+                        'tax_ids': [fields.Command.clear()],
+                        'name':'SCONTO' if general_discount < 0 else 'MAGGIORAZIONE',
+                        'sequence': sequence,
+                        'price_unit': general_discount
+                    })
```
Comment (smetl) [inline addons/l10n_it_edi/models/account_edi_format.py:731]:
```suggestion
                    invoice_form.invoice_line_ids = [Command.create({
                        'move_id': invoice_form.id,
                        'tax_ids': [fields.Command.clear()],
                        'name':'SCONTO' if general_discount < 0 else 'MAGGIORAZIONE',
                        'sequence': sequence,
                        'price_unit': general_discount
                    })]
```

File: addons/l10n_it_edi/tests/test_edi_import.py [hunk 3]
```diff
@@ -129,3 +131,11 @@ def test_receive_same_vendor_bill_twice(self):
...
         invoices = self.env['account.move'].search([('payment_reference', '=', 'TWICE_TEST')])
         self.assertEqual(len(invoices), 1)
+
+    def test_discount_line(self):
+        """Test discount line addition after bill import"""
+
+        content = etree.fromstring(self.invoice_discount_content)
+        invoices = self.edi_format._create_invoice_from_xml_tree(self.invoice_discount_file, content)
+
+        self.assertEqual(invoices.line_ids.name, 'SCONTO')
```
Comment (jco-odoo) [inline addons/l10n_it_edi/tests/test_edi_import.py:141]: If we do the entire test, we could check if the line values are correct entirely...

File: addons/l10n_it_edi/models/account_edi_format.py (hunk 2 above)
Comment (adda-odoo) [reply addons/l10n_it_edi/models/account_edi_format.py:731]: Hey @smetl! My initial commit was in this format but @ThomasBeckers suggested using `create`.

**Comment by jco-odoo** [comment]
Taken care of in https://github.com/odoo/odoo/pull/121010

**Pull request closed without merging**